core/src/test/java/com/linecorp/armeria/client/HttpResponseDecoderTest.java

```diff
@@ -0,0 +1,128 @@
+import org.junit.runners.Parameterized.Parameters;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.google.common.collect.ImmutableList;
+
+import com.linecorp.armeria.client.HttpResponseDecoder.HttpResponseWrapper;
+import com.linecorp.armeria.client.retry.Backoff;
+import com.linecorp.armeria.client.retry.RetryStrategy;
+import com.linecorp.armeria.client.retry.RetryingHttpClientBuilder;
+import com.linecorp.armeria.common.HttpMethod;
+import com.linecorp.armeria.common.HttpRequest;
+import com.linecorp.armeria.common.HttpResponse;
+import com.linecorp.armeria.common.SessionProtocol;
+import com.linecorp.armeria.common.logging.RequestLogAvailability;
+import com.linecorp.armeria.server.ServerBuilder;
+import com.linecorp.armeria.testing.server.ServerRule;
+
+@RunWith(Parameterized.class)
+public class HttpResponseDecoderTest {
+    private static final Logger logger = LoggerFactory.getLogger(HttpResponseDecoderTest.class);
+
+    @ClassRule
+    public static ServerRule rule = new ServerRule() {
+        @Override
+        protected void configure(ServerBuilder sb) throws Exception {
+            sb.service("/", (ctx, req) -> HttpResponse.of("Hello, Armeria!"));
+        }
+    };
+
+    @Parameters(name = "{index}: protocol={0}")
+    public static Collection<SessionProtocol> protocols() {
+        return ImmutableList.of(H1C, H2C);
+    }
+
+    private final SessionProtocol protocol;
+
+    public HttpResponseDecoderTest(SessionProtocol protocol) {
+        this.protocol = protocol;
+    }
+
+    /**
+     * This test would be passed because the {@code cancelAction} method of the {@link HttpResponseWrapper} is
+     * invoked in the event loop of the {@link Channel}.
+     */
+    @Test
+    public void confirmResponseStartAndEndInTheSameThread() throws InterruptedException {
+        final AtomicBoolean failed = new AtomicBoolean();
+        final RetryStrategy strategy =
+                (ctx, cause) -> CompletableFuture.completedFuture(Backoff.withoutDelay());
+        final HttpClient client = new HttpClientBuilder(protocol.uriText() + "://127.0.0.1:" + rule.httpPort())
+                // This increases the execution duration of 'endResponse0' of the DefaultRequestLog,
+                // which means that we have more chance to reproduce the bug if two threads are racing
+                // for notifying RESPONSE_END to listeners.
+                .contentPreview(100)
+                // In order to use a different thread to send a request.
+                .decorator(new RetryingHttpClientBuilder(strategy).maxTotalAttempts(2).newDecorator())
+                .decorator((delegate, ctx, req) -> {
+                    final AtomicReference<Thread> responseStartedThread = new AtomicReference<>();
+                    ctx.log().addListener(log -> {
+                        responseStartedThread.set(Thread.currentThread());
+                        if (!ctx.eventLoop().inEventLoop()) {
```

Comment:
I think this can cause this test failure when a connection is not established by the VM conditions or etc.(In that case, this is executed by the thread who calls `execute()`.)
So I think it's perfectly fine to remove this and only check the thread is same or not. (Still, in that case, the thread which starts end ends the response will be the same.)
WDYT?

Reply:
Thanks! I misunderstood what you explained offline before. 👍